core-primitives/stf-executor/src/executor.rs

```diff
@@ -0,0 +1,440 @@
+/*
+	Copyright 2021 Integritee AG and Supercomputing Systems AG
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+		http://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+
+*/
+
+// #[cfg(all(not(feature = "std"), feature = "sgx"))]
+// use crate::sgx_reexport_prelude::*;
+
+#[cfg(all(not(feature = "std"), feature = "sgx"))]
+use std::untrusted::time::SystemTimeEx;
+
+use crate::traits::StfExecuteTimedGettersBatch;
+use crate::{
+	error::{Error, Result},
+	traits::{
+		StatePostProcessing, StfExecuteGenericUpdate, StfExecuteShieldFunds,
+		StfExecuteTimedCallsBatch, StfExecuteTrustedCall, StfUpdateState,
+	},
+	ExecutedOperation, ExecutionHashes, ExecutionResult, ExecutionStatus,
+};
+use codec::{Decode, Encode};
+use ita_stf::{
+	hash::TrustedOperationOrHash,
+	stf_sgx::{shards_key_hash, storage_hashes_to_update_per_shard},
+	AccountId, ShardIdentifier, StateTypeDiff, Stf, TrustedCall, TrustedCallSigned,
+	TrustedGetterSigned,
+};
+use itp_ocall_api::{EnclaveAttestationOCallApi, EnclaveOnChainOCallApi};
+use itp_stf_state_handler::handle_state::HandleState;
+use itp_storage::StorageEntryVerified;
+use itp_storage_verifier::GetStorageVerified;
+use itp_types::{Amount, OpaqueCall, H256};
+use log::*;
+use sgx_externalities::{SgxExternalities, SgxExternalitiesTrait};
+use sp_runtime::{
+	app_crypto::sp_core::blake2_256,
+	traits::{Block as BlockT, Header, UniqueSaturatedInto},
+};
+use std::{
+	collections::HashMap,
+	fmt::Debug,
+	format,
+	result::Result as StdResult,
+	sync::Arc,
+	time::{Duration, SystemTime},
+	vec::Vec,
+};
+
+/// STF Executor implementation
+///
+///
+pub struct StfExecutor<OCallApi, StateHandler> {
+	ocall_api: Arc<OCallApi>,
+	state_handler: Arc<StateHandler>,
+}
```

Comment:
This file now contains all the relevant implementation for executing calls, getters and operations in general on the STF. It's probably a bit large and does not have any unit-tests. There are a couple of indirect tests of these parts, in the `enclave-runtime`.
Unfortunately this module is only available in `sgx` feature, because we heavily use `Stf` and all related structures.

Reply:
Having unit tests would certainly be a great improvement. And we're lacking them everywhere. So I'm not sure an issue 'create unit tests for this' will do much good. In my experience those kind of issues never get prioritized in a sprint 😝 Next time we touch this code and we extend or change the behavior, we need to introduce unit tests.

Reply:
You got me there.. can't reason against that ;-(

Reply:
Well you could argue to still do them now 😄 I'm a bit on the fence about it myself, because I like to write unit tests. And on the other hand, I should make some progress on the actual feature (#301) I'm trying to achieve

Reply:
I also agree. Maybe add a //FIXME comment. Not an issue but atleast something that someone will stumble upon when this code part is touched?